File: app/src/main/java/de/christinecoenen/code/zapp/app/mediathek/api/request/QueryRequest.java
package de.christinecoenen.code.zapp.app.mediathek.api.request;


import android.text.TextUtils;

import androidx.annotation.NonNull;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;

@SuppressWarnings({"UnusedReturnValue"})
public class QueryRequest implements Serializable {

	@SuppressWarnings("FieldCanBeLocal")
	private final String[] ALLOWED_CHANNELS = new String[]{
		"ARD",
		"ZDF",
		"SWR",
		"NDR",
		"BR",
		"ARTE.DE",
		"DW",
		"3Sat",
		"SRF",
		"MDR",
		"SR",
		"KiKA",
		"HR",
		"RBB",
		"ORF",
		"WDR",
		"rbtv",
		"PHOENIX",
		"ZDF-tivi",
		"SRF.Podcast"
	};

	private final List<Query> queries = new ArrayList<>();
	private final List<Query> alowedChannelsQueries = new ArrayList<>();

	private String sortBy = "timestamp";
	private String sortOrder = "desc";

	@SuppressWarnings("FieldCanBeLocal")
	private final boolean future = false;
	private int offset = 0;
	private int size = 30;

	public QueryRequest() {
		for (String allowedChannel : ALLOWED_CHANNELS) {
			this.alowedChannelsQueries.add(new Query(allowedChannel, "channel"));
		}

		ResetQueries();
	}

	public QueryRequest setSimpleSearch(String queryString) {
		ResetQueries();

		if (!TextUtils.isEmpty(queryString)) {
			this.queries.add(new Query(queryString, "title", "topic"));
		}
		return this;
	}

	public QueryRequest addQuery(String fieldName, String queryString) {
		this.queries.add(new Query(queryString, fieldName));
		return this;
	}

	public QueryRequest setSortBy(String sortBy) {
		this.sortBy = sortBy;
		return this;
	}

	public QueryRequest setSortOrder(String sortOrder) {
		this.sortOrder = sortOrder;
		return this;
	}

	public QueryRequest setOffset(int offset) {
		this.offset = offset;
		return this;
	}

	public QueryRequest setSize(int size) {
		this.size = size;
		return this;
	}

	private void ResetQueries() {
		this.queries.clear();
		this.queries.addAll(alowedChannelsQueries);
	}

	@NonNull
	@Override
	public String toString() {
		return "QueryRequest{" +
			"queries=" + queries +
			", sortBy='" + sortBy + '\'' +
			", sortOrder='" + sortOrder + '\'' +
			", future=" + future +
			", offset=" + offset +
			", size=" + size +
			'}';
	}
}

File: app/src/main/java/de/christinecoenen/code/zapp/app/mediathek/api/request/QueryRequest.kt
package de.christinecoenen.code.zapp.app.mediathek.api.request

import androidx.annotation.Keep
import com.google.gson.annotations.SerializedName
import java.io.Serializable

@Keep
class QueryRequest : Serializable {

	private val sortBy: String = "timestamp"
	private val sortOrder: String = "desc"

	var future: Boolean = true
	var offset: Int = 0
	var size: Int = 30

	@SerializedName("duration_min")
	var minDurationSeconds: Int = 0

	@SerializedName("duration_max")
	var maxDurationSeconds: Int? = null

	private val queries: MutableList<Query> = mutableListOf()

	@Transient
	private val channels: MutableSet<MediathekChannel> = mutableSetOf()

	@Transient
	private var queryStrings: MutableSet<String> = mutableSetOf()

	init {
		resetQueries()
	}

	fun setChannels(channels: List<MediathekChannel>): QueryRequest = apply {
		this.channels.clear()
		this.channels.addAll(channels)

		resetQueries()
	}

	fun setQueryString(queryString: String): QueryRequest = apply {
		this.queryStrings.clear()
		this.queryStrings.add(queryString)

		resetQueries()
	}

	fun setQueryStrings(queryStrings: List<String>): QueryRequest = apply {
		this.queryStrings.clear()
		this.queryStrings.addAll(queryStrings)

		resetQueries()
	}

	private fun resetQueries() {
		queries.clear()

		// set search query
		queries.addAll(queryStrings.filter { it.isNotEmpty() }.map { queryString ->
			Query(queryString, "title", "topic")
		})

		// We do not allow an empty channel Filter as the result would always be empty.
		// Instead we filter for all available channels. This also excludes all channels
		// not defined in MediathekChannel enum (like ARTE.FR).
		//Timber.d(channels.size.toString())
		if (channels.isEmpty()) {
			channels.addAll(MediathekChannel.entries.toTypedArray())
		}

		// set all currently allowed channels
		for (allowedChannel in channels) {
			queries.add(Query(allowedChannel.apiId, "channel"))
		}
	}

	fun isEmpty() = queries.isEmpty()

	override fun toString(): String {
		return "QueryRequest(sortBy='$sortBy', sortOrder='$sortOrder', future=$future, offset=$offset, size=$size, queries=$queries)"
	}
}
